src/backend/distributed/commands/utility_hook.c

```diff
@@ -517,13 +517,32 @@ multi_ProcessUtility(PlannedStmt *pstmt,
 	{
 		PostStandardProcessUtility(parsetree);
 
+		/*
+		 * We call MarkInvalidateForeignKeyGraph in preprocess without knowing
+		 * that if command would fail or not. However, we don't want to invalidate
+		 * foreign key graph for failed DDL commands, so here we don't call
+		 * InvalidateForeignKeyGraphForDDL.
+		 * On the other hand, we should still reset shouldInvalidateForeignKeyGraph
+		 * flag as next DDL command still relies on this flag to invalidate foreign
+		 * key graph.
+		 */
+		shouldInvalidateForeignKeyGraph = false;
+
 		PG_RE_THROW();
 	}
 	PG_END_TRY();
 
 	/*
 	 * Post process for ddl statements
 	 */
+
```

Comment:
why do we change the place of this code?

Reply:
I indeed did this change in 51ab536. Previously, we were setting foreign key graph to be invalidated at `PostprocessCreateTable` but we were executing `InvalidateFKeyCacheForDDL` before `PostprocessCreateTable`, so we were actually not invalidating foreign key graph for create table commands

Reply:
I think the way to  tackle it is to call `InvalidateForeignKeyGraph()` in  `PostprocessCreateTable ()` ? 

My understanding is that for commands in `PostProcess` we are good to directly call `InvalidateForeignKeyGraph()`. For `PreProcess`, we mark as we should invalidate, and do the invalidation after the command is executed via process utility, as we get the information from catalog tables.

Does that make sense?